Batch Mode

as-cii opened the pull request:
With this PR, when passing `{batchMode: true}` to `Patch` constructor, a [splay tree](https://en.wikipedia.org/wiki/Splay_tree) will be used as the underlying data structure for managing changes.

#### Benchmarks

![screen shot 2016-02-04 at 11 40 13](https://cloud.githubusercontent.com/assets/482957/12812745/16c257a2-cb34-11e5-9a96-1a69ef527f5b.png)

_(Benchmarks generated using https://gist.github.com/as-cii/3f020dd5789f9c649bc8. Copy the content of that gist and then execute `pbpaste | git apply && npm run tdd` in your terminal)_

As you can notice, the two implementations have similar performance except for the case of non-overlapping sequential insertions, where the splay tree is ~ 3x faster. 

Another case where the splay tree outperforms the treap is when repeatedly splicing in a ordered manner (e.g. splice one character in each row, splice one more character after the previous one in the same rows, etc.). By calling `balance` on the patch after each mass insertion, we manage to make the splay tree ~ `125ms` faster.

### Manual Testing

In practice, when inserting a single character on ~ 3000 lines, the improvement is even more noticeable.

|        |                                                                                                                                            |
|--------|--------------------------------------------------------------------------------------------------------------------------------------------|
| Before | ![screen shot 2016-02-03 at 04 57 50](https://cloud.githubusercontent.com/assets/482957/12772848/e9faf106-ca35-11e5-9ca5-b5010949312a.png) |
| After  |  ![screen shot 2016-02-03 at 04 57 31](https://cloud.githubusercontent.com/assets/482957/12772847/e9c540e2-ca35-11e5-9dd3-fdbedbafb894.png) |

Splay trees in this case are 8x faster (also, please note that the total time showed here is the cost of maintaining two patches, as I ran the tests with https://github.com/atom/text-buffer/pull/127).

When inserting multiple characters, splicing into those two patches is only 2x faster, but I think it's because of the `stoppedChangingPatch` (which may be acting similarly to the automated benchmark above, "Repeated Sequential Non-Overlapping Insertions"), so DSW should help here as well.

Overall, I am pretty happy with the results and I'd love to hear what you guys think about this. Also, before merging this PR, I'd like to experiment with a couple of things:

* [x] Implement a `balance` function that uses DSW and re-assess performance.

/cc: @maxbrunsfeld @nathansobo @kuychaco for :eyes: :racehorse: :racehorse: :racehorse:

maxbrunsfeld commented:
:clap: :tada: Amazing!

> Use splay tree in MarkerIndex as well

This is a great idea; find-and-replace's main bottleneck is sequential on-overlapping insertions.

kuychaco commented:
:sparkles: So excited about this! Nice work @as-cii!

as-cii commented:
So, I have introduced the `rebalance()` method and updated the benchmarks accordingly. As we predicted, rebalancing the patch causes a `130ms` speedup for repeated sequential non-overlapping insertions.

I believe we can drop the treap implementation and always use splay trees + on-demand rebalancing.

@nathansobo @maxbrunsfeld @kuychaco: what do you think? It'd be awesome to have another pair of :eyes: on the code before merging it. :pray: 

In the meantime, I am going to work on implementing the same structure on `MarkerIndex` and check if it makes things better in there as well. :+1:

nathansobo commented on src/patch.js:
```diff
@@ -32,15 +65,16 @@ export default class Patch {
-    this.bubbleNodeUp(startNode)
-    endNode.priority = -2
-    this.bubbleNodeUp(endNode)
+    if (endNode.left !== startNode) {
```
Could you just change the order in which the nodes are splayed?

as-cii replied:
I am not sure we can, because we want `endNode` to be at the top of the tree. When splaying that node, however, a `zig-zig` could take place and, as a result, `startNode` would become the left child of `endNode`'s initial parent (i.e. becoming the left child of `endNode`'s left child after the splay) and that's basically what that if clause is preventing.

![zig-zig](https://cloud.githubusercontent.com/assets/482957/12840153/dee11a28-cbe0-11e5-8269-3098b099b033.gif)

_(`10` would be the `startNode` in this case and `30` the `endNode`)_

Am I missing something? Perhaps we could write a comment to make that clearer?

nathansobo commented on src/patch.js:
```diff
@@ -19,6 +14,44 @@ export default class Patch {
     return new Iterator(this)
   }
 
+  rebalance () {
```
Another thing I was wondering is could we rebalance only a portion of the tree after a splice sequence... Maybe from the leftmost new node to the rightmost new node. Would shrink N dramatically in many cases but might be tricky to implement.

nathansobo commented on src/patch.js:
```diff
@@ -19,6 +14,44 @@ export default class Patch {
+    this.createBalancedTree()
+  }
+
+  createBackbone () {
```
Another presentation of the algorithm named this "vine" instead of "backbone" which I think is a bit nicer. Pretty subjective obviously but thought I would mention it.

as-cii replied:
:+1:

nathansobo commented on src/patch.js:
```diff
@@ -216,6 +231,42 @@ export default class Patch {
     return rightAncestor
   }
 
+  splayNode (node) {
```
👌🏻 looking legit

nathansobo commented:
We may want to measure outliers on these benchmarks in addition to averaging, because the splay tree trades performance of individual operations for throughput. I want to be sure we understand the worst case time for individual operations before we commit to splay trees in interactive workloads (the display layer patch). Would love to hear your thoughts and analysis on the interactive use case... Do you think we could apply DSW on "hot" ranges of the tree only, like following sequential insertions? For the change aggregation case this is a pretty clear "go" for me. For the synchronous change aggregation use case we don't even have to rebalance the tree because we only care about the overall time for building and traversing the tree.

as-cii commented:
> For the change aggregation case this is a pretty clear "go" for me. For the synchronous change aggregation use case we don't even have to rebalance the tree because we only care about the overall time for building and traversing the tree.

Agreed, we only need rebalance for the patch that maintains `onDidStopChanging` changes. :+1: 

> We may want to measure outliers on these benchmarks in addition to averaging, because the splay tree trades performance of individual operations for throughput. I want to be sure we understand the worst case time for individual operations before we commit to splay trees in interactive workloads (the display layer patch). Would love to hear your thoughts and analysis on the interactive use case...

So I thought more about this and I think that a bad scenario would occur when we have lots of splices that didn't happen because of mass multi-cursor editing, but were the result of e.g. the user repeatedly pasting a line. As a result, when editing the very first pasted line, a noticeable amount of time will pass because the tree is hugely unbalanced. On average, the splay tree will actually perform as fast (or even faster) as the treap, but editing that line will feel really slow. In this case, it's also probably ineffective to rebalance only a portion of the tree, as the whole tree might be unbalanced (in the case of repeatedly pasting into an empty file).
I have simulated the above scenario with the following code (a user pasting 3000 lines, one by one):

```js
// pasting lots of lines
let treapPatch = new Patch({combineChanges: true, batchMode: false})
let splayPatch = new Patch({combineChanges: true, batchMode: true})
let random = new Random(Date.now())
for (var row = 0; row < 3000; row++) {
  for (var j = 0; j < 10; j++) {
    let start = {row: row, column: random.intBetween(0, 200)}
    let oldExtent = {row: 0, column: random.intBetween(0, 5)}
    let newExtent = {row: 0, column: random.intBetween(0, 5)}
    treapPatch.splice(start, oldExtent, newExtent)
    splayPatch.splice(start, oldExtent, newExtent)
  }
}

// changing only the very first pasted row
treapTime += measure(() => treapPatch.spliceInput({row: 0, column: 0}, {row: 1, column: 0}, {row: 1, column: 0}))
splayTime += measure(() => splayPatch.spliceInput({row: 0, column: 0}, {row: 1, column: 0}, {row: 1, column: 0}))
for (var j = 0; j < 10; j++) {
  let start = {row: 0, column: random.intBetween(0, 200)}
  let oldExtent = {row: 0, column: random.intBetween(0, 5)}
  let newExtent = {row: 0, column: random.intBetween(0, 5)}
  treapTime += measure(() => treapPatch.splice(start, oldExtent, newExtent))
  splayTime += measure(() => splayPatch.splice(start, oldExtent, newExtent))
}
```

And, while the treap takes `~ 1ms` to access the first row, the splay tree takes `~ 60ms`.

This probably means that treaps would be more appropriate in this case, as they pay something more when touching nodes but also constantly keep the tree in a good shape.

However, and as an alternative, we could put the `Patch` in `batchMode` for `DisplayLayer` as well when we know that a mass editing operation is going to happen (e.g. when we compute the whole display layer the first time, or when the user edits on lots of cursors), then rebalance and put the `Patch` back to `interactiveMode` at the end (we could even make this happen automatically, as part of the switch from batch to interactive mode). Rebalancing in this case will also have to assign non-random, yet valid priorities to nodes in order to respect the treap properties.

Thoughts?

maxbrunsfeld commented:
> And, while the treap takes ~ 1ms to access the first row, the splay tree takes ~ 60ms. This probably means that treaps would be more appropriate in this case

How long does rebalancing the splay tree take in this scenario? Could we just keep a count of how many separate un-ordered edits we have made, and explicitly rebalance the tree once it exceeds a certain value and the user stops editing for a few hundred milliseconds?

as-cii commented:
> How long does rebalancing the splay tree take in this scenario? Could we just keep a count of how many separate un-ordered edits we have made, and explicitly rebalance the tree once it exceeds a certain value and the user stops editing for a few hundred milliseconds?

For that particular scenario it takes `~ 40ms`, but what would happen if the user scrolled to the top of the file before those few hundred milliseconds elapse? Even just scrolling might be enough to perceive the slowness, as we'd need to translate input/output positions to render them on screen (because of line numbers).

I feel like for interactive workloads the treap might be just fine, as I am concerned that those heuristics may not take into account some corner case. And we still might go back and forth from and to batch mode in `DisplayLayer` when some big chunk of work comes in and we're sure that splay trees + rebalancing are the right/performant choice.

What do you think? :thought_balloon:

maxbrunsfeld commented:
> What do you think? :thought_balloon:

I think that if it's simple to switch a given Patch back and forth between treap and splay modes, then we should just do it. But if it adds much complexity, we should try to reproduce the scenario above in a real editor, and consider whether it's really matters before adding the complexity.

nathansobo commented:
My instinct is that the worst imbalance in the display layer scenario would occur after long runs of sequential splices, which is why I thought that only paying to rebalance those specific nodes might be cheaper. But to me it's seeming like we should maybe just go for using splay trees with explicit rebalancing at obvious moments and see how we do in practice. We have way worse slowdowns elsewhere in the system and we can always experiment with more specific optimization scenarios later. Considering the display layer isn't even on master yet, it seems like we have the information we need to proceed with optimizations based on batching changes and worry about the interactive case later.

as-cii commented:
> My instinct is that the worst imbalance in the display layer scenario would occur after long runs of sequential splices, which is why I thought that only paying to rebalance those specific nodes might be cheaper.

Agreed, although my concern is that there could be operations which create a pretty large tree (i.e. the copy-paste operation described above) where rebalancing only part of the tree could not be effective.

> But to me it's seeming like we should maybe just go for using splay trees with explicit rebalancing at obvious moments and see how we do in practice. We have way worse slowdowns elsewhere in the system and we can always experiment with more specific optimization scenarios later. Considering the display layer isn't even on master yet, it seems like we have the information we need to proceed with optimizations based on batching changes and worry about the interactive case later.

Sounds good, I have still reverted 4acb4a6 so that we can fall back to using the treap implementation when we're implementing the `DisplayLayer`. We can always delete that later after assessing the performance on practical use cases.

Seems like this is ready to go now, so I am going to :shipit: and we can always refine it later if we need to. Thanks everyone for the thorough and thoughtful feedback! :sparkles:
